src/mcp/shared/session.py

```diff
@@ -214,19 +229,44 @@ async def send_request(
         result_type: type[ReceiveResultT],
         request_read_timeout_seconds: timedelta | None = None,
         metadata: MessageMetadata = None,
+        progress_callback: ProgressFnT | None = None,
     ) -> ReceiveResultT:
         """
         Sends a request and wait for a response. Raises an McpError if the
         response contains an error. If a request read timeout is provided, it
         will take precedence over the session read timeout.
 
+        If progress_callback is provided any progress notifications sent from the
+        receiver will be passed back to the sender
+
         Do not use this method to emit notifications! Use send_notification()
         instead.
         """
 
         request_id = self._request_id
         self._request_id = request_id + 1
 
+        progress_id = None
+        send_request = None
+
+        if progress_callback is not None:
```

Comment:
Nasty if/else branching, ugly but functional main challenge is how to inject meta into params when the type of the request is unknown. Open to ideas on how to make this cleaner